### ISSUE TITLE
Dataset Creation

### ISSUE THREAD

**Issue by ming783** [asafkar/deep_shadow#5]
could you kindly share some details about the dataset creation process for deep_shadow？

**Comment by asafkar** [comment]
The synthetic dataset was created in Blender, using various 3d objects, with a constant camera position and varying lights.
I extracted the shadows using a layer which does so in Blender.

**Comment by ming783** [comment]
Thank you for your reply. I would also like to ask if normal maps and depth files can be directly exported from Blender? Does each line in "all_object_lights.txt" represent the position of a light source, and do the values in each line correspond to the x, y, and z coordinates respectively?

**Comment by asafkar** [comment]
Yes - both normal and depth maps are possible outputs of Blender - take care that you may need to normalize the depth between 0 and 1, and the normal maps should be rotated in the right direction (it's easiest just to place the camera at [0,0,1] and then you won't have any problems).

regarding all_object_lights.txt - what you wrote is correct.

**Comment by ming783** [comment]
If my object is very large, for example, a sphere with a diameter of 500 meters, will increasing the lighting range significantly affect the training process?

**Comment by asafkar** [comment]
There are a few assumptions that were made, one of which, is that the light source is distant enough. 
If you have a large object, you can place the lights far enough (I would start with at least 5x distance of the furthest point of the object from it's center); or you can also just scale the object and keep the lights closer.

**Comment by ming783** [comment]
Thank you for your reply. Your response has been very helpful to me.

**Comment by ming783** [comment]
"Hello, I would like to ask if the light source must be inside the view frustum and cannot be outside of it?

**Comment by asafkar** [comment]
Hi, 
The light source can be outside the view frustum, it shouldn't make a difference

**Comment by ming783** [comment]
D:\Anaconda3\envs\DS\lib\site-packages\torch\optim\lr_scheduler.py:28: UserWarning: The verbose parameter is deprecated. Please use get_last_lr() to access the learning rate.
  warnings.warn("The verbose parameter is deprecated. Please use get_last_lr() "
light src tensor([8, 0]) in frame
light src tensor([98,  6]) in frame
light src tensor([98,  6]) in frame
light src tensor([98,  6]) in frame
light src tensor([8, 0]) in frame
light src tensor([48,  0]) in frame
light src tensor([48,  0]) in frame
light src tensor([48,  0]) in frame
epoch 0 running loss = 700848.3125。The main problem I am facing now is that the training loss is too high and difficult to converge. I’m not sure about the exact reason, but I suspect it might be related to my normal maps or depth maps, which I exported directly from Blender.

**Comment by asafkar** [comment]
I suppose you need to run for more iterations. I don't recall which losses I encountered when training, but you need many more epochs before it will converge.

**Comment by ming783** [comment]
light src tensor([8, 0]) in frame
light src tensor([98, 6]) in frame
light src tensor([98, 6]) in frame
light src tensor([98, 6]) in frame
light src tensor([8, 0]) in frame
light src tensor([48, 0]) in frame
light src tensor([48, 0]) in frame
light src tensor([48, 0]) in frame，Is this kind of display also normal? Because I didn't encounter these issues when training on your dataset

**Comment by asafkar** [comment]
short answer - it's OK.

In order to execute the viewshed, the code needs to generate point coordinates from the light source to the pixel at the border - 
https://github.com/asafkar/deep_shadow/blob/2550db7d338e1af6164c56c7c7892ad56584d448/viewshed.py#L220 

There is a check whether the light source is projected into the frame of the generated shadow image or outside of it, and the line is generated from the intersection of the light source to the point with the border, or just from the light source if it's the frame.

**Comment by ming783** [comment]
Normed Depth Prediction Error = nan;Does this issue during training indicate that there might be a problem with my depth maps?

**Comment by asafkar** [comment]
That shouldn't be the case. 
I suggest first trying to general a simple object with 4 far point-light sources, and making sure that works, before continuing. 

You are welcome to upload these and I will try to have a look.

**Comment by ming783** [comment]
Could you please help me take a look at my dataset?
















At 2024-11-28 22:18:23, "Asaf Karnieli" ***@***.***> wrote:

That shouldn't be the case.
I suggest first trying to general a simple object with 4 far point-light sources, and making sure that works, before continuing.

You are welcome to upload these and I will try to have a look.

**Comment by ming783** [comment]
Hello, thank you very much for your help. I have successfully created the dataset and completed depth estimation. However, I have a question regarding the model training. Why does the model only perform well on the dataset used during training? I thought there are fundamental physical relationships between shadow positions, light sources, and the 3D structure of objects, so the model should exhibit some level of generalization across different scenes. I find this point a bit puzzling.
